infra/event-sync/src/main/java/com/simprints/infra/eventsync/sync/up/tasks/EventUpSyncTask.kt

```diff
@@ -74,115 +78,104 @@ internal class EventUpSyncTask @Inject constructor(
-     * events while we upload the current one), but given the relatively small amount of data to
-     * upload, and how critical this system is, we are happy to trade off speed for reliability
-     * (through simplicity and low resource usage)
-     */
     private fun uploadEvents(
         projectId: String,
-        canSyncAllDataToSimprints: Boolean,
-        canSyncBiometricDataToSimprints: Boolean,
-        canSyncAnalyticsDataToSimprints: Boolean
-    ): Flow<Int> = flow {
+        config: ProjectConfiguration,
+    ) = flow {
         Simber.d("[EVENT_REPO] Uploading")
-
-        eventRepository.getAllClosedSessionIds(projectId).forEach { sessionId ->
-            // The events will include the SessionCaptureEvent event
-            Simber.d("[EVENT_REPO] Uploading session $sessionId")
-            try {
-                eventRepository.getEventsFromSession(sessionId).let {
-                    attemptEventUpload(
-                        it, projectId,
-                        canSyncAllDataToSimprints,
-                        canSyncBiometricDataToSimprints,
-                        canSyncAnalyticsDataToSimprints
-                    )
-                    this.emit(it.size)
-                }
-            } catch (ex: Exception) {
-                if (ex is JsonParseException || ex is JsonMappingException) {
-                    attemptInvalidEventUpload(projectId, sessionId)?.let { this.emit(it) }
-                } else {
-                    Simber.i("Failed to un-marshal events for $sessionId")
+        try {
+            val sessionScopes = eventRepository.getAllClosedSessions(projectId).associateWith {
+                try {
+                    eventRepository.getEventsFromSession(it.id)
+                        .also { listOfEvents -> emit(listOfEvents.size) }
+                } catch (ex: JsonParseException) {
+                    Simber.i("Failed to un-marshal events")
                     Simber.e(ex)
+                    null
                 }
             }
-        }
-    }
 
-    private suspend fun attemptEventUpload(
-        events: List<Event>, projectId: String,
-        canSyncAllDataToSimprints: Boolean,
-        canSyncBiometricDataToSimprints: Boolean,
-        canSyncAnalyticsDataToSimprints: Boolean
-    ) {
-        try {
-            val filteredEvents = when {
-                canSyncAllDataToSimprints -> {
-                    events
-                }
-                canSyncBiometricDataToSimprints -> {
-                    events.filter {
-                        it is EnrolmentEventV2 || it is PersonCreationEvent || it is FingerprintCaptureBiometricsEvent || it is FaceCaptureBiometricsEvent
-                    }
+            val corruptedScopes = sessionScopes.filterValues { it == null }.keys
+
+            // Re-emitting the number of uploaded corrupted events
+            attemptInvalidEventUpload(projectId, corruptedScopes).collect { emit(it) }
```

Comment:
It is still happening; it is just slightly rearranged. 
The previous version was unintuitive. The JSON check did not happen "on request" but rather on database read (and since those were pretty much 1-to-1 with requests, it looked like it was related to request). 
The new version still does the same thing, just grouping the corrupted and uncorrupted scope uploads into separate processes.